internal/reminder/reminder.go

```diff
@@ -149,70 +132,111 @@ func (r *reminder) Stop() {
+	for _, repo := range repos {
+		logger.Debug().Str("repo", repo.ID.String()).
+			Time("previously", repo.ReminderLastSent.Time).Msg("updating reminder_last_sent")
+		err := r.store.UpdateReminderLastSentById(ctx, repo.ID)
+		if err != nil {
+			logger.Error().Err(err).Str("repo", repo.ID.String()).Msg("unable to update reminder_last_sent")
+			return []error{err}
+		}
+		// TODO: Send the actual reminders
 	}
 
-	return os.WriteFile(r.cfg.CursorFile, buf.Bytes(), 0600)
+	return nil
 }
 
-// restoreCursorState restores the cursor state from a file
-// Not thread-safe, should be called from a single goroutine
-func (r *reminder) restoreCursorState(ctx context.Context) error {
+func (r *reminder) getRepositoryBatch(ctx context.Context) ([]db.Repository, error) {
 	logger := zerolog.Ctx(ctx)
-	logger.Debug().Msg("restoring cursor state")
 
-	if _, err := os.Stat(r.cfg.CursorFile); os.IsNotExist(err) {
-		return nil
+	logger.Debug().Msgf("fetching repositories after cursor: %s", r.repositoryCursor)
+	repos, err := r.store.ListRepositoriesAfterID(ctx, db.ListRepositoriesAfterIDParams{
+		ID:    r.repositoryCursor,
+		Limit: int64(r.cfg.RecurrenceConfig.BatchSize),
+	})
+	if err != nil {
+		return nil, err
 	}
 
-	data, err := os.ReadFile(r.cfg.CursorFile)
+	eligibleRepos, err := r.getEligibleRepositories(ctx, repos)
 	if err != nil {
-		return err
+		return nil, err
 	}
+	logger.Debug().Msgf("%d/%d repositories are eligible for reminders", len(eligibleRepos), len(repos))
 
-	buf := bytes.NewBuffer(data)
-	dec := gob.NewDecoder(buf)
+	r.updateRepositoryCursor(ctx, repos)
 
-	cursorData := make(map[string]interface{})
+	return eligibleRepos, nil
+}
+
+func (r *reminder) getEligibleRepositories(ctx context.Context, repos []db.Repository) ([]db.Repository, error) {
+	eligibleRepos := make([]db.Repository, 0, len(repos))
 
-	if err := dec.Decode(&cursorData); err != nil {
-		return err
+	repoIds := make([]uuid.UUID, 0, len(repos))
+	for _, repo := range repos {
+		repoIds = append(repoIds, repo.ID)
 	}
 
-	if val, ok := cursorData["projectListCursor"]; ok {
-		if v, ok := val.(string); ok {
-			r.projectListCursor = v
-		} else {
-			return fmt.Errorf("projectListCursor is not a string")
-		}
+	oldestRuleEvals, err := r.store.ListOldestRuleEvaluationsByRepositoryId(ctx, repoIds)
+	if err != nil {
+		return nil, err
 	}
 
-	if val, ok := cursorData["repoListCursor"]; ok {
-		if v, ok := val.(map[projectProviderPair]string); ok {
-			r.repoListCursor = v
-		} else {
-			return fmt.Errorf("repoListCursor is not a map[projectProviderPair]string")
+	idToLastUpdatedMap := make(map[uuid.UUID]time.Time)
+	for _, oldestRuleEval := range oldestRuleEvals {
+		idToLastUpdatedMap[oldestRuleEval.RepositoryID] = oldestRuleEval.OldestLastUpdated
+	}
+
+	for _, repo := range repos {
+		if oldestRuleEval, ok := idToLastUpdatedMap[repo.ID]; ok &&
+			oldestRuleEval.Add(r.cfg.RecurrenceConfig.MinElapsed).Before(time.Now()) {
```

Comment:
Why do this two-step loop here?  It feels like either passing in the cutoff to the query (i.e. `SELECT ... WHERE rde.last_updated < $1 AND re.repository_id = ANY(...) GROUP BY re.repository_id` or the equivalent `HAVING`), or looping through the results once should be sufficient:

```golang
cutoff := time.Now().Sub(r.cfg.RecurrenceConfig.MinElapsed)
for _, evalTime := range oldestRuleEvals {
  if evalTime.OldestLastUpdated.Before(cutoff) {
    eligibleRepos = append(eligibleRepos, repo)
  }
}
```

Reply:
`ListOldestRuleEvaluationsByRepositoryId` only queries `rule_evaluations` and `rule_details_eval`, so it returns a slice of:
```go
type ListOldestRuleEvaluationsByRepositoryIdRow struct {
	RepositoryID      uuid.UUID `json:"repository_id"`
	OldestLastUpdated time.Time `json:"oldest_last_updated"`
}
```

We have to iterate over repos as the repo object isn't returned.

Reply:
Ah, that seems like it's worth a comment about how we're doing a bunch of transforms of types to fit into the sqlc-generated code (rather than for some other reason, like performance or thread-safety).

Also, a slight preference for:

```golang
cutoff := time.Now().Sub(r.cfg.RecurrenceConfig.MinElapsed)
for _, repo := range repos {
  if t, ok := idToLastUpdate[repo.ID]; ok && t.Before(cutoff) {
    ....
  }
}
```

This has two benefits:

1. You can fit the condition on one line (shortening the time var that lives for a single line, removing `Map` from the name of the map, and doing the date math on a separate line).
2. You only do the date math _once_, and only fetch `time.Now()` once, rather than using a slightly different time for each check.